[3.x] Fix browser and engine detection of web client

richard67 opened the pull request:
Pull Request for Issue #112 

### Summary of Changes

This pull request (PR) fixes diverse bugs in the browser and engine detection of the web client.

First of all it adds a check if a version part was found in the `$userAgent` parameter at 2 places in the "detectEngine" method of web client in order not to run into a PHP error "Undefined array key 1" when a known user agent is used without a version part.

See the linked issue and another CMS issue https://github.com/joomla/joomla-cms/issues/44469 .

In addition, it adds a check for user agent = "Chrome Privacy Preserving Prefetch Proxy" to the "detectBrowser" method to make sure that no wrong version is detected, and it adds a unit test case for that user agent and for user agents without a version where we would have run into the "Undefined array key 1" error.

The user agent "Chrome Privacy Preserving Prefetch Proxy" is used by the https://developer.chrome.com/blog/private-prefetch-proxy and does not provide any further information like e.g. a version.

The Privacy Preserving Prefetch Proxy is a feature which Google gradually roll out on their Chrome browsers. It allows the prefetching of cross-origin content without exposing user information to the destination website.

That means when you see previews of websites in search results, your user agent will not be exposed to the website where this preview comes from until you navigate to that website. That's the case where the PHP warnings mentioned in the referred issue and CMS issues happen, which is fixed by this PR here.

In addition, this PR fixes the wrong check `$version[0] === 537.36` for "AppleWebKit/537.36" which should be `$version[0] === '537.36'` in the "detectEngine" method and adds a test case for that because existing test cases for real user agents did not cover that case.

Last but not least, this PR fixes the wrong check `$version[0] < 28` for Google Chrome, which works well with the existing test case for version "54.0.2840.71" but not for a current version > 99 like e.g. "131.0.0.0". The fix is to use version_compare, which works regardless of how many dots you have in a version, and this PR adds a new test data set for that.

Finally, this PR adds a text case for the issue #114 fixed with PR #131 .

### Testing Instructions

#### Prepare testing environment

On a development environment (Linux computer or Windows with WSL, with composer and a git client), clone the joomla-framework/application repository into a local folder and prepare for running unit tests on the default branch `3.x-dev`:

```
mkdir test-pr-132
cd test-pr-132
git clone https://github.com/joomla-framework/application.git .
composer update --prefer-stable
```

Check that unit tests succeed on your environment:
```
vendor/bin/phpunit --testdox
```
Result:
```
OK (213 tests, 373 assertions)
```

#### Add new unit tests from this PR and check failing test cases

Now apply only the changes in file `Tests/Web/WebClientTest.php` from this PR, e.g. by downloading the file from the raw view on GiHub:
https://raw.githubusercontent.com/joomla-framework/application/df5fc53c3e15817c6449e372ed415e9ab95530b9/Tests/Web/WebClientTest.php

Do **_not_** apply the changes from file `src/Web/WebClient.php` from this PR.

Run unit tests:
```
vendor/bin/phpunit --testdox
```
Results:
- Detect browser fails for the new data set no. 42 with user agent "Chrome Privacy Preserving Prefetch Proxy" because it detects a wrong version "Privacy".
```
 ✘ Detect browser with data set #42
   │
   │ Version detection failed
   │ Failed asserting that two strings are equal.
   │ --- Expected
   │ +++ Actual
   │ @@ @@
   │ -'Privacy'
   │ +''
   │
   │ /home/richard/GitHub/joomla-framework/application/Tests/Web/WebClientTest.php:525
   │
```
- Detect engine fails for the new data set no. 16 with the user agent of a Google Chrome version 131.0.0.0 because of the wrong check for version < 28, so it detects engine `WebClient::WEBKIT` instead of `WebClient::BLINK`.
```
 ✘ Detect engine with data set #16
   │
   │ Engine detection failed.
   │ Failed asserting that 24 matches expected 12.
   │
   │ /home/richard/GitHub/joomla-framework/application/Tests/Web/WebClientTest.php:589
   │
```
- Detect engine fails for the new data set no. 42 with user agent "Chrome Privacy Preserving Prefetch Proxy" because it detects engine `WebClient::WEBKIT` instead of `WebClient::BLINK`.
```
 ✘ Detect engine with data set #42
   │
   │ Engine detection failed.
   │ Failed asserting that 24 matches expected 12.
   │
   │ /home/richard/GitHub/joomla-framework/application/Tests/Web/WebClientTest.php:589
   │
```
- Detect engine fails for the new data set no. 43 with user agent "Chrome" without any version because because it detects engine `WebClient::WEBKIT` instead of `WebClient::BLINK`.
```
 ✘ Detect engine with data set #43
   │
   │ Engine detection failed.
   │ Failed asserting that 24 matches expected 12.
   │
   │ /home/richard/GitHub/joomla-framework/application/Tests/Web/WebClientTest.php:589
   │
```
- Detect engine fails for the new data set no. 44 with user agent "AppleWebKit" without any version with a PHP notice.
```
 ⚠ Detect engine with data set #44
```
- Detect engine fails for the new data set no. 45 with user agent "'AppleWebKit/537.36" because it detects engine `WebClient::WEBKIT` instead of `WebClient::BLINK` due to the wrong check `$version[0] === 537.36` which should be `$version[0] === '537.36'`.
```
 ✘ Detect engine with data set #45
   │
   │ Engine detection failed.
   │ Failed asserting that 24 matches expected 12.
   │
   │ /home/richard/GitHub/joomla-framework/application/Tests/Web/WebClientTest.php:589
   │
```
- Overall result:
```
FAILURES!
Tests: 231, Assertions: 403, Failures: 5, Warnings: 2.
Remaining self deprecation notices (3)
```

#### Run unit tests on the code changes from this PR

Now apply also the changes from file `src/Web/WebClient.php` from this PR, e.g. by downloading the file from the raw view on GiHub:
https://raw.githubusercontent.com/joomla-framework/application/df5fc53c3e15817c6449e372ed415e9ab95530b9/src/Web/WebClient.php

Run unit tests:
```
vendor/bin/phpunit --testdox
```
Result:
```
OK (231 tests, 403 assertions)
```

### Documentation Changes Required

None as far as I know.

richard67 committed:
Fix undef. array key for user agent w/o version

Quy commented on src/Web/WebClient.php:
```diff
@@ -395,19 +395,19 @@ protected function detectEngine($userAgent)
         } elseif (\stripos($userAgent, 'Chrome') !== false) {
             $result  = \explode('/', \stristr($userAgent, 'Chrome'));
-            $version = \explode(' ', $result[1]);
+            $version = \count($result) > 1 ? \explode(' ', $result[1]) : false;
```
A simpler fix?
```
if (isset($result[1])) {
}
```

richard67 replied:
Would also work. But is that really simpler?

richard67 replied:
Other question I just ask myself is if this PR here is right at all. The purpose of the method is to detect the client rendering engine, but that "Chrome Privacy Preserving Prefetch Proxy" is not really a client but some kind of a proxy, it seems. I have to investigate further.

Quy replied:
Less code to read/process..one test vs a count and an extra test for false.
```
            $result  = \explode('/', \stristr($userAgent, 'Chrome'));
            if (isset($result[1])) {
            	$version = \explode(' ', $result[1]);

            	if ($version[0] >= 28) {
                	$this->engine = self::BLINK;
            	} else {
                	$this->engine = self::WEBKIT;
            	}
            }
```

richard67 replied:
But this code behaves different to what I proposed. In your case $this->engine will not be set at all when there was no version.

richard67 replied:
In my proposal it is set to BLINK. Not sure what is right, but it’s different.

richard67 replied:
I‘ve tried to make it similar to other checks for versions a few lines below.

richard67 replied:
But maybe your proposal is right and there should not be any engine detected when there is no version. I have to investigate that, or maybe someone can advise.

Quy replied:
You are absolutely right. 👍 Please ignore my noise.

Maybe this?
```
$version = \count($result) > 1 ? \explode(' ', $result[1]) : false;
$version = isset($result[1]) ? \explode(' ', $result[1]) : false;
```

richard67 replied:
Yes, that would simplify my code. Will apply it in a minute. But still need to investigate if my fix is right. It may be next weekend until I can do that.

Quy replied:
It feels off to me too.

richard67 committed:
Use isset instead of count

Quy commented on src/Web/WebClient.php:
```diff
@@ -395,19 +395,19 @@ protected function detectEngine($userAgent)
+            $version = isset($result[1]) ? \explode(' ', $result[1]) : false;
 
-            if ($version[0] >= 28) {
+            if ($version === false || $version[0] >= 28) {
```
Maybe it should be the same as AppleWebKit? I assume `self::WEBKIT` to be the fallback value.
`$version !== false &&`

richard67 replied:
I would assume an actual version if no version is given. "Chrome Privacy Preserving Prefetch Proxy" is a real user agent used for the https://developer.chrome.com/blog/private-prefetch-proxy feature of Google Chrome.

richard67 replied:
I've added a unit test, and it fails with a reason. The detectBrowser method would need to be adjusted to not detect a version="Privacy " when the user agent is "Chrome Privacy Preserving Prefetch Proxy"

richard67 replied:
Changed to draft status until I have checked further.

richard67 replied:
I thin it is OK to use `if ($version === false || $version[0] >= 28) {` here as the https://developer.chrome.com/blog/private-prefetch-proxy has been added to Chrome 103 for Android, and for other cases without version we can also assume Chrome version >= 28.

ssabatini replied:
Looks like this should work. (I'm currently not able to test this, sorry.)

richard67 replied:
I've refactored the code so it works the same way but is better readable at this place here.

richard67 committed:
Add test data for web client tests

Add unit test data for user agent "Chrome Privacy Preserving Prefetch Proxy" for web client.

richard67 committed:
Don't detect version='Privacy'

Don't detect version='Privacy' in detectBrowser method of web client.

richard67 committed:
Check version='Privacy' only if browser is Chrome

Quy commented on src/Web/WebClient.php:
```diff
@@ -395,19 +395,19 @@ protected function detectEngine($userAgent)
-                if ($version[0] === 537.36) {
+                if ($version !== false && $version[0] === 537.36) {
                     // AppleWebKit/537.36 is Blink engine specific, exception is Blink emulated IEMobile, Trident or Edge
                     $this->engine = self::BLINK;
```
Not related to your PR, but this doesn't seem right. The assignment on line 412 is replaced on line 417.

richard67 replied:
Yes ... weird. Would have to check with Git Blame if someone has removed a "return" statement or how else that code was changed so it is that weird.

richard67 replied:
It was added with this PR in 2016: #62 . And that created that nonsense.

Quy replied:
Move line 417 above the if statement (line 406) should fix the issue.

ssabatini replied:
or put it into an else clause:
```PHP
} elseif (\stripos($userAgent, 'AppleWebKit') !== false || \stripos($userAgent, 'blackberry') !== false) {
  if (\stripos($userAgent, 'AppleWebKit') !== false) {
     ...
  }
  else
  {
    // Evidently blackberry uses WebKit and doesn't necessarily report it.  Bad RIM.
    $this->engine = self::WEBKIT;
  }
```

richard67 replied:
I've refactored the code so it makes sense. It seems the current unit tests do not really cover the case for AppleWebKit/537.36 because they pass regardless if this code is present or not. It also doesn't matter for the existing unit test cases if I check for `$version[0] === 537.36` (which is wrong) or `$version[0] === '537.36'`. I will fix that soon with another commit.

richard67 committed:
Early check for prefetch proxy in detectBrowser

richard67 committed:
Refactor version checks in detectEngine

richard67 committed:
Add unit test case for issue #114 / PR #131

richard67 committed:
Add unit test cases for user agents w/o version

richard67 committed:
Fix wrong check for AppleWebKit/537.36 + test case

richard67 commented:
@Quy Thanks so far for your review comments and suggestions. I've refactored my code so it matches to what you had suggested first and is much better readable, and I have fixed another issue with a wrong type safe comparison and added unit tests to test my changes. See the updated description and testing instructions. I'd be really happy if you could review my PR again and test it as described with help of the unit tests.

Quy commented on src/Web/WebClient.php:
```diff
@@ -287,6 +287,14 @@ protected function detectBrowser($userAgent)
+        // Check for Google's Private Prefetch Proxy
+        if ($userAgent === 'Chrome Privacy Preserving Prefetch Proxy') {
+            // Private Prefetch Proxy does not provide any further details like e.g. version
+            $this->browser  = self::CHROME;
```
```suggestion
            $this->browser = self::CHROME;
```

richard67 committed:
CS

Co-authored-by: Quy <quy@nomonkeybiz.com>

Quy commented on src/Web/WebClient.php:
```diff
@@ -287,6 +287,14 @@ protected function detectBrowser($userAgent)
+            return;
+        }
+
         $patternBrowser = '';
```
```suggestion
```
This is assigned when there is match otherwise it is an unnecessary assignment.

richard67 replied:
Yes it is assigned, but only when there is a match. There is no `else` so it might end unassigned. Then later below in line 333 is it used to create the pattern:
```
$pattern = '#(?<browser>Version|' . $patternBrowser . ')[/ :]+(?<version>[0-9.|a-zA-Z.]*)#';
```
And this also works when the `$patternBrowser` is an empty string.

I would prefer not to change that.

Quy replied:
It won't reach/execute line 333 without `$this->browser`. Only with a match that `$this->browser` and `$patternBrowser` will be assigned. I understand your decision. It can be done in a separate PR.

richard67 replied:
Yes, you are right. But I prefer a separate PR.

richard67 committed:
Fix wrong check for version < 28 and add test case
